### PR TITLE
Wraps stasis behind procs like blindness

### PR DESCRIPTION
**Pull request by JJRcop**
## About The Pull Request
Stasis is now controlled with the `stasis` and `cure_stasis` procs, like `become_blind` and `cure_blind`. Like blindness, these procs add/remove a new stasis trait, and handle adding and removing the status effect itself.

~~I also moved the `can_defib` proc out of defib to carbon, and made defibrillation nanites use that so they are synced with defibs.~~

## Why It's Good For The Game
Using the status effect on its own was not very compatible with multiple sources.

`Life` now takes advantage of `HAS_TRAIT` to check for stasis, which may have performance improvements over the `has_status_effect` proc.

## Changelog
:cl: JJRcop
refactor: Stasis now uses traits, which support multiple sources.
/:cl:

### CODE REVIEW

**Comment by itseasytosee** [comment]
Stasis gun when?

**Comment by trollbreeder** [comment]
Maybe we can tie extremely low temperature to stasis.
Then we can finally rework cryo to be long term patient storage instead of being a crutch like cloning.
Just ideas.

**Comment by JJRcop** [comment]
Split off can_defib change to separate branch

**Comment by optimumtact** [comment]
>adding a trait 

ok I can see it

>wrapping a status effect in some useless helper procs

I don't get it

**Comment by optimumtact** [comment]
in fact you could just add a multiple source framework to status effects as well.

**Comment by JJRcop** [comment]
I'd like to get the code out of a status effect if possible

File: code/game/machinery/stasis.dm [hunk 1]
```diff
@@ -144,9 +144,9 @@
 		return
 	var/mob/living/L_occupant = occupant
 	if(stasis_running())
-		if(!IS_IN_STASIS(L_occupant))
+		if(!HAS_TRAIT(L_occupant, TRAIT_STASIS))
```
Comment (XDTM) [inline code/game/machinery/stasis.dm]: make it `HAS_TRAIT_FROM` `STASIS_MACHINE_TRAIT`, in case someone got stasis from other sources

**Comment by XDTM** [comment]
Ending metabolism is simple enough, bbut consantly pushing forward time of death requires a ticking system, and traits are static by design. Unless we rework time of death to be a counter instead of a timestamp, a status effect might still be necessary.

File: code/game/machinery/stasis.dm (hunk 1 above)
Comment (JJRcop) [reply code/game/machinery/stasis.dm]: good catch

**Comment by JJRcop** [comment]
@optimumtact I don't see an issue with this, but you would rather not have a trait and add multiple sources to status effects?

Most status effects don't need this, so I would instead make a status effect subtype with multiple sources.

**Comment by ninjanomnom** [comment]
You could add a signal in ADD_TRAIT for when a trait is added and do something like

```
#define COMSIG_DATUM_TRAIT_ADDED(trait) "trait_added_[trait]"
...
[inside ADD_TRAIT()]
  SEND_SIGNAL(target, COMSIG_DATUM_TRAIT_ADDED(trait), source)
```

**Comment by XDTM** [comment]
I suppose if you do it you might as well make it generic and make it
```
#define COMSIG_DATUM_TRAIT_ADDED "trait_added"
...
[inside ADD_TRAIT()]
  SEND_SIGNAL(target, COMSIG_DATUM_TRAIT_ADDED, trait, source, cumulative)
```

Where `cumulative` being TRUE means that the trait was already present and this is just a new source, in case you need to react to a particular source being added.

**Comment by ninjanomnom** [comment]
I can't imagine why someone would want to know when any trait is added, that's why I had the trait as part of the signal so you could register for only the signal for the trait you want

**Comment by XDTM** [comment]
Ah, my bad, i didn't see it was a macro

**Comment by JJRcop** [comment]
@ninjanomnom Wouldn't we want this as a global signal with the target and source as an argument?

**Comment by XDTM** [comment]
The way in his example has the benefit of not triggering the registered proc on every single trait addition, and only doing it on the relevant one.

**Comment by JJRcop** [comment]
I was talking about `SEND_GLOBAL_SIGNAL`, which sends the signal on SSdcs, so instead of having to subscribe to every mob, you can subscribe to SSdcs and know when anyone gets a certain status effect

**Comment by JJRcop** [comment]
Made a new status effect with multiple sources in #49237

**Pull request closed without merging**